[lldb] Swift OS plugin

felipepiovezan opened the pull request:
This PR implements the required functionality for a swift OS plugins that converts Tasks into Threads, allowing LLDB to transparently work with Tasks as if they were threads.
In particular, this addresses _all_ (famous last words?) the outstanding stepping issues in swift async code.

felipepiovezan committed:
[lldb][swift] Swift OS Plugin

This commit introduces an OS plugin converting Process threads into
OperatingSystem threads that represent a Task. This is done by
inspecting ThreadLocalStorage for a specific location where the swift
runtime writes information about the current Task.

With this plugin, ThreadPlans now work properly in async code: a step
action on a Thread is now effectively an action on a Task, and LLDB will
correctly track that. ThreadPlans are effectively "Task Plans".

adrian-prantl commented on lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp:
```diff
@@ -0,0 +1,164 @@
+//===-- OperatingSystemSwift.cpp -----------------------------------------===//
```
Not important. "OperatingSystemSwift" sounds slightly wrong to me. What about "OperatingSystemSwiftTasks" or "OperatingSystemSwiftConcurrency"?

medismailben replied:
Yeah the name seems misleading the full lldb affordance name is "Operating System Thread Plugin"

felipepiovezan replied:
> the full lldb affordance name is "Operating System Thread Plugin"

While this may be true, it's worth noting that the only other plugin also uses the short version: "OperatingSystemX", where  X = Python.


> OperatingSystemSwiftTasks

I'm going to rename to this!

adrian-prantl commented on lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp:
```diff
@@ -0,0 +1,164 @@
+    return nullptr;
+
+  /// Heuristic to only load the plugin for swift programs.
+  FileSpec swift_module_name = FileSpec("libswiftCore.dylib");
```
This only works in a very narrow subset of Swift programs. You should defer to SwiftLanguageRuntime's helper code.
The cases this doesn't support are: non-Darwin (".dylib"), statically linked Swift stdlib, embedded Swift (no dylibs at all).

felipepiovezan replied:
Do you have a helper function mind? The only thing I found is `GetLikelySwiftImageNamesForModule`, which has some very weird behavior I'm not sure I understand.

jimingham replied:
You could also arrange that this can only be accessed by the SwiftLanguageRuntime, and then its loading will be gated on the Runtime loading, which already has to do this sort of computation.

felipepiovezan replied:
@jimingham I looked into that option, but AFAICT the SwiftLanguageRuntime is loaded regardless of modules.
I.e. the first call to "load runtimes" will always load it, even if the relevant libraries haven't been loaded yet

felipepiovezan replied:
@adrian-prantl I don't think the code in SwiftLanguageRuntime handles the cases you mentioned either, it always returns a dylib for Darwin platforms:

```
static ConstString GetStandardLibraryName(Process &process) {
  // This result needs to be stored in the constructor.
  PlatformSP platform_sp(process.GetTarget().GetPlatform());
  if (platform_sp)
    return platform_sp->GetFullNameForDylib(
        ConstString(SwiftLanguageRuntime::GetStandardLibraryBaseName()));
  return {};
}

ConstString SwiftLanguageRuntime::GetStandardLibraryName() {
  return ::GetStandardLibraryName(*m_process);
}

ConstString PlatformDarwin::GetFullNameForDylib(ConstString basename) {
  if (basename.IsEmpty())
    return basename;

  StreamString stream;
  stream.Printf("lib%s.dylib", basename.GetCString());
  return ConstString(stream.GetString());
}
```

felipepiovezan replied:
I found a private method in there that maybe we could expose...

adrian-prantl commented on lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp:
```diff
@@ -0,0 +1,164 @@
+      swift::tls_get_key(swift::tls_key::concurrency_task) * ptr_size;
+  // Offset of a Task ID inside a Task data structure, guaranteed by the ABI.
+  // See Job in swift/RemoteInspection/RuntimeInternals.h.
+  m_task_id_offset = 4 * ptr_size + 4;
```
You should check that the swift concurrency version is still 1 here.

See: https://github.com/swiftlang/swift/pull/76437

felipepiovezan commented on lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    // Mask higher bits to avoid conflicts with core thread IDs.
+    uint64_t masked_task_id = 0xdeadbeef00000000 | *task_id;
```
Btw it is probably fine to not do this, TIDs are generally higher numbers and there might be value in exposing the task id directly. If we do mask, we should use some other mask

jimingham commented on lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    // Mask higher bits to avoid conflicts with core thread IDs.
+    uint64_t masked_task_id = 0xdeadbeef00000000 | *task_id;
```
This will be the visible "TID" for running tasks, right?  And what I'd have to type if I wanted to set a "task specific" breakpoint?  If so, we might not want it to be deadbeef.  That's cute if you know the reference, but otherwise somewhat off-putting...

felipepiovezan replied:
Oh yeah, I'm actually looking for suggestions here, if we should mask it at all and what to use if we do
Deadbeef was just something to help me grep the logs as I debugged.

jimingham commented on lldb/source/Plugins/OperatingSystem/Swift/OperatingSystemSwift.cpp:
```diff
@@ -0,0 +1,164 @@
+    ThreadSP swift_thread = [&]() -> ThreadSP {
+      // If we already had a thread for this Task in the last stop, re-use it.
+      if (ThreadSP old_thread = old_thread_list.FindThreadByID(masked_task_id);
+          IsOperatingSystemPluginThread(old_thread))
```
Do you have to handle the case where when you stopped the first time, this Task was backed by "core thread 1" but the next time you stopped it was backed by a different core thread?  I don't see where you are changing the backing thread in this case?

felipepiovezan replied:
> I don't see where you are changing the backing thread in this case?

It's just a few lines below, both cases are handled the same way:

```
    swift_thread->SetBackingThread(real_thread);
```

jimingham commented:
I note that the only swift specific bits of this are finding the task on a particular thread, and a few checks about whether this should be used that I think belong more properly in the LanguageRuntime that owns the OS plugin...

Would it be possible to make an `OperatingSystemTaskOverlay` that has all the generic algorithms, and then a swift version that just adds the "how to find the task ID"?.  The concept is pretty general, and we could probably mock it with an artificial OS plugin so we could test a lot of this w/o having to involve Swift...

felipepiovezan committed:
[lldb][swift] Add helper function to find Swift concurrency's module version

felipepiovezan commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeNames.cpp:
```diff
@@ -302,6 +302,7 @@ class ThreadPlanRunToAddressOnAsyncCtx : public ThreadPlan {
     auto &target = thread.GetProcess()->GetTarget();
     m_funclet_bp = target.CreateBreakpoint(destination_addr, true, false);
     m_funclet_bp->SetBreakpointKind("async-run-to-funclet");
+    m_funclet_bp->SetThreadID(thread.GetID());
```
to refresh reviewer's memories: this is the thread plan to "run through a task switch and stop when we schedule this same ***async function*** again. This plan now sets thread specific breakpoints, since they are now task breakpoints!

If we somehow push this type of plan without having the OS plugin, we are at a risk of breaking in the wrong task. But if the plugin is not there and there are tasks, we are going to have much bigger problems anyway. For example, this code has been associated with a lot of LLDB crashes because it triggers stack plan migration (this is a downstream-only thing that becomes dead code with this patch, and I'll delete it in a follow up PR)

medismailben replied:
Should be this an inline comment in the source file ?

felipepiovezan replied:
I would argue that in this case we don't need any comments, because this is just the "natural" thing to do. If anything, it is the _old_ implementation that deserved a comment, as it was essentially making all threads stop.

kastiglione replied:
> This plan now sets thread specific breakpoints, since they are now task breakpoints!

I'm not sure I understand what you're saying with this.

felipepiovezan replied:
I just noticed that the writing was indeed confusing, my bad.
The important thing to note here is that this plan is no longer special, it should work like any other plan: when it sets a breakpoint, it sets a thread specific breakpoint (otherwise it would cause all threads to stop).

felipepiovezan replied:
A follow up patch will remove this plan in favor of the RunToBreakpointPlan

felipepiovezan committed:
[lldb] Add experimental target option controlling swift OS plugin

This is here mostly out of an abundance of caution, the default is to
use the plugins.

medismailben commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -91,6 +91,10 @@ const char *SwiftLanguageRuntime::GetStandardLibraryBaseName() {
   return "swiftCore";
 }
 
+const char *SwiftLanguageRuntime::GetConcurrencyLibraryBaseName() {
```
nit: Can this return a `StringRef` or `StringLiteral` ?

felipepiovezan replied:
It could, for sure, but it would break with the pattern of the other functions above. We can do a cleanup and change all of them in a separate PR

medismailben commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -131,6 +149,11 @@ static bool IsStaticSwiftRuntime(Module &image) {
 }
 
+static bool IsStaticSwiftConcurrency(Module &image) {
+  static const ConstString task_switch_symbol("_swift_task_switch");
```
This could be:
```suggestion
  static constexpr llvm::StringLiteral task_switch_symbol = "_swift_task_switch";
```

felipepiovezan replied:
The rationale here was also just to keep uniformity with the other identical helper functions in this file. But also the call `image.FindFirstSymbolWithNameAndType` requires a ConstString, so if we have a StringLiteral we would be instantiating a ConstString on every call to this method, as opposed to just once.

medismailben commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -168,6 +191,52 @@ static ModuleSP findRuntime(Process &process, RuntimeKind runtime_kind) {
   return runtime_image;
 }
 
+ModuleSP SwiftLanguageRuntime::findConcurrencyModule(Process &process) {
```
nit: casing

```suggestion
ModuleSP SwiftLanguageRuntime::FindConcurrencyModule(Process &process) {
```

medismailben commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -168,6 +191,52 @@ static ModuleSP findRuntime(Process &process, RuntimeKind runtime_kind) {
+}
+
+std::optional<uint32_t>
+SwiftLanguageRuntime::findConcurrencyDebugVersion(Process &process) {
```
ditto

felipepiovezan replied:
oops, will change!

medismailben commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.h:
```diff
@@ -98,6 +98,16 @@ class SwiftLanguageRuntime : public LanguageRuntime {
   }
+
+  /// Returns the Module containing the Swift Concurrency runtime, if it exists.
+  static lldb::ModuleSP findConcurrencyModule(Process &process);
```
ditto

medismailben commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.h:
```diff
@@ -98,6 +98,16 @@ class SwiftLanguageRuntime : public LanguageRuntime {
+  /// returned.
+  /// Returns 0 for versions of the module prior to the introduction
+  /// of versioning.
+  static std::optional<uint32_t> findConcurrencyDebugVersion(Process &process);
```
ditto

medismailben commented on lldb/source/Plugins/OperatingSystem/SwiftTasks/OperatingSystemSwiftTasks.cpp:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    // Mask higher bits to avoid conflicts with core thread IDs.
+    uint64_t masked_task_id = 0x0000000f00000000 | *task_id;
```
Nice hack :p

felipepiovezan replied:
To be honest I don't think we need to / should do this, because conflicts are handled without issues... it is just  a way to visually know we have a fake thread.

kastiglione replied:
how are the conflicts handled?

felipepiovezan replied:
It's a couple of lines down, on line 100

medismailben commented on lldb/source/Plugins/OperatingSystem/SwiftTasks/OperatingSystemSwiftTasks.cpp:
```diff
@@ -0,0 +1,155 @@
+
+    ThreadSP swift_thread = [&]() -> ThreadSP {
+      // If we already had a thread for this Task in the last stop, re-use it.
+      if (ThreadSP old_thread = old_thread_list.FindThreadByID(masked_task_id);
+          IsOperatingSystemPluginThread(old_thread))
```
First time I'm seeing this syntax :p Don't you want to run the second part of the if condition only if the first one if valid ?
```suggestion
      if ((ThreadSP old_thread = old_thread_list.FindThreadByID(masked_task_id)) && IsOperatingSystemPluginThread(old_thread))
```

felipepiovezan replied:
`IsOperatingSystemPluginThread` is designed to take a pointer argument, so it knows how to handle nullptrs.

medismailben reviewed (commented):
cool stuff! I left some comments / suggestions :)

felipepiovezan commented:
addressed review comments

kastiglione commented on lldb/source/Plugins/OperatingSystem/SwiftTasks/OperatingSystemSwiftTasks.cpp:
```diff
@@ -0,0 +1,155 @@
+  size_t ptr_size = process.GetAddressByteSize();
+  // Offset of a Task ID inside a Task data structure, guaranteed by the ABI.
+  // See Job in swift/RemoteInspection/RuntimeInternals.h.
+  m_task_id_offset = 4 * ptr_size + 4;
```
is this the job id? if so should the name reflect that? also if so, do we know if we can ignore the task's upper 32 bit identifier?

felipepiovezan replied:
Based on what I saw in the runtime code, I think it's safe to do so. It is what is being done in other tools anyway.

I'll rename this to "job id"!

kastiglione commented on lldb/source/Plugins/OperatingSystem/SwiftTasks/OperatingSystemSwiftTasks.cpp:
```diff
@@ -0,0 +1,155 @@
+    // Mask higher bits to avoid conflicts with core thread IDs.
+    uint64_t masked_task_id = 0x0000000f00000000 | *task_id;
+
+    ThreadSP swift_thread = [&]() -> ThreadSP {
+      // If we already had a thread for this Task in the last stop, re-use it.
+      if (ThreadSP old_thread = old_thread_list.FindThreadByID(masked_task_id);
+          IsOperatingSystemPluginThread(old_thread))
+        return old_thread;
+
+      std::string name = llvm::formatv("Swift Task {0:x}", *task_id);
+      llvm::StringRef queue_name = "";
+      return std::make_shared<ThreadMemory>(*m_process, masked_task_id, name,
+                                            queue_name,
+                                            /*register_data_addr*/ 0);
+    }();
```
The lambda through me off. Is there something I'm missing that it can't be an if/else?

```suggestion
    ThreadSP swift_thread;
    // If we already had a thread for this Task in the last stop, re-use it.
    if (ThreadSP old_thread = old_thread_list.FindThreadByID(masked_task_id);
        IsOperatingSystemPluginThread(old_thread)) {
      swift_thread = old_thread;
    } else {
      std::string name = llvm::formatv("Swift Task {0:x}", *task_id);
      llvm::StringRef queue_name = "";
      swift_thread = std::make_shared<ThreadMemory>(*m_process, masked_task_id, name,
                                                    queue_name,
                                                    /*register_data_addr*/ 0);
    }
```

felipepiovezan replied:
The missing piece is that this creates an uninitialized variable :) 
I'll move this code to a helper function

jimingham commented on lldb/source/Plugins/OperatingSystem/SwiftTasks/OperatingSystemSwiftTasks.cpp:
```diff
@@ -0,0 +1,155 @@
+
+  if (*concurrency_version <= 1)
+    return new OperatingSystemSwiftTasks(*process);
+  LLDB_LOGF(log,
```
Did you intend to only log the version in the failure case?

jimingham replied:
Also, if you've encountered a more recent version of the concurrency library that this lldb recognizes, we should tell the user that.  Could there be a `Status` path threaded to  this `Initialize`.

felipepiovezan replied:
Yes, I was originally only logging the problematic case, but there is no harm in logging it all the time, so I'll move the log statement before the conditional

felipepiovezan replied:
>Also, if you've encountered a more recent version of the concurrency library that this lldb recognizes, we should tell the user that. Could there be a Status path threaded to this Initialize.

Is this how we do user warnings? I was grepping around and found `Debugger::ReportWarning`, with a matching `std::once_flag` flag. If this is the standard way of doing things, and to make sure a warning for this plugin is only reported once, it seems like the implementation would be:

1. call ReportWarning inside `OperatingSystemSwiftTasks` directly,
2. have that class own the std::once flag. (Note that the "CreateInstance" methods are called many times, e.g. every time libraries are loaded)

jimingham replied:
I wasn't reading this right.  There really isn't anything that can fail here, since we're just looking for a module by name, and then the version will only be an empty optional if we didn't find the module.  And not finding the module is fine, we just might not be using swift concurrency...  So there's no warning to issue here, IIUC.

jimingham replied:
What you've added here looks right.  We only warn if we found the concurrence library but it looks funny to us.

jimingham commented on lldb/source/Plugins/OperatingSystem/SwiftTasks/OperatingSystemSwiftTasks.cpp:
```diff
@@ -0,0 +1,155 @@
+        return old_thread;
+
+      std::string name = llvm::formatv("Swift Task {0:x}", *task_id);
+      llvm::StringRef queue_name = "";
```
If the core thread had a queue name, would we want to use it here?

felipepiovezan replied:
Just to register an offline conversation, an empty queue name will make ThreadMemory propagate the backing thread queue name once this fix gets merged: https://github.com/llvm/llvm-project/pull/126128

felipepiovezan commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntime.cpp:
```diff
@@ -2795,7 +2795,11 @@ llvm::Expected<lldb::addr_t> GetTaskAddrFromThreadLocalStorage(Thread &thread) {
 #else
```
Just realized that I never wrote a commit message for this after I moved it out of draft state. Will do so shortly

jimingham reviewed (approved):
This LGTM.  This worked more easily than I expected it would, that's great.

I don't see why the tests shouldn't also work on Linux/Windows but we don't need to hold up this patch to resolve that.

medismailben reviewed (approved):
LGTM!

felipepiovezan commented:
@swift-ci test

felipepiovezan commented:
Update the commit message

felipepiovezan commented:
@swift-ci test

felipepiovezan commented:
@swift-ci test

felipepiovezan commented:
Rebased

felipepiovezan commented:
@swift-ci test
